packages/router/utils.ts

```diff
@@ -959,23 +965,70 @@ export class DeferredData {
     this.subscriber?.(true);
   }
 
+  async resolveData(signal: AbortSignal) {
+    let aborted = false;
+    if (!this.done) {
+      let onAbort = () => this.cancel();
+      signal.addEventListener("abort", onAbort);
+      aborted = await new Promise((resolve) => {
+        this.subscribe((aborted) => {
+          signal.removeEventListener("abort", onAbort);
+          if (aborted || this.done) {
+            resolve(aborted);
+          }
+        });
+      });
+    }
+    return aborted;
+  }
+
   get done() {
     return this.pendingKeys.size === 0;
   }
+
+  get unwrappedData() {
+    invariant(
+      this.data !== null && this.done,
+      "Can only unwrap data on initialized and settled deferreds"
+    );
+
+    if (this.data instanceof Promise) {
+      return unwrapDeferredPromise(this.data);
+    }
+
+    if (Array.isArray(this.data)) {
+      return this.data.map((value) => unwrapDeferredPromise(value));
+    }
+
+    if (typeof this.data === "object") {
+      return Object.entries(this.data).reduce(
+        (acc, [key, value]) =>
+          Object.assign(acc, {
+            [key]: unwrapDeferredPromise(value),
+          }),
+        {}
+      );
+    }
+
+    throw new Error("Invalid DeferredData type");
+  }
 }
 
-/**
- * @private
- * Utility class we use to hold deferred promise rejection values
- */
-export class DeferredError extends Error {}
+function isDeferredPromise(value: any): value is DeferredPromise {
+  return (
+    value instanceof Promise && (value as DeferredPromise)._tracked === true
+  );
+}
 
-/**
- * Check if the given error is a DeferredError generated from a deferred()
- * promise rejection
- */
-export function isDeferredError(e: any): e is DeferredError {
-  return e instanceof DeferredError;
+function unwrapDeferredPromise(value: any) {
+  if (!isDeferredPromise(value)) {
+    return value;
+  }
+
+  if (value._error) {
+    throw value._error;
```

Comment:
Need to double check on this - we want rejected deferred values on fetchers to trigger error boundaries just like a rejected `fetcher.load` would